### PR TITLE
Crawler Module refactored

### PR DESCRIPTION
**Pull request by maciejkocur**
Interfaces changed to have more testable code.
Annotated as Functional.
Change lamdas to methode references if possible.
Handle null existing by filtering in streams operations.

### CODE REVIEW

**Comment by maciejkocur** [comment]
@danone771909 , Please review our code.

**Comment by ghost** [comment]
Now it looks better, than previous implementation, but still it would be nice to change `Collection<String>` to `List<String>`. This would help other developers using this API to choose correct collection, I mean where the order does not matter and adding elements is fast (the most important point). :ghost:

File: src/main/java/io/github/bookcrawler/core/BookExtractor.java [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 import java.util.Collection;
 
+@FunctionalInterface
```
Comment (KPP13) [inline src/main/java/io/github/bookcrawler/core/BookExtractor.java:7]: :+1:

File: src/main/java/io/github/bookcrawler/core/BookStore.java [hunk 2]
```diff
@@ -20,17 +20,17 @@ public String domainUrl() {
...
         @Override
         public BooksLinkCrawler crawler() {
-            return new EmpikBookLinksCrawler(new JsoupSourceScrapper());
+            return new EmpikBookLinksCrawler();
         }
 
         @Override
-        public EmpikBookInfoParser parser() {
-            return new EmpikBookInfoParser(new JsoupSourceScrapper());
+        public BookInfoParser parser() {
```
Comment (KPP13) [inline src/main/java/io/github/bookcrawler/core/BookStore.java:32]: nice change to the base type, you use polymorphism now :+1:

File: src/main/java/io/github/bookcrawler/core/impl/EmpikBooksExtractor.java [hunk 3]
```diff
@@ -1,16 +1,29 @@
...
+    private SourceScrapper sourceScrapper;
+
+    public EmpikBooksExtractor(SourceScrapper sourceScrapper) {
+        this.sourceScrapper = sourceScrapper;
+    }
+
+    @Override
     public Collection<BookInfo> extract() {
-        return EMPIK.parser().parse(EMPIK.crawler().crawl());
+        return EMPIK.crawler().crawl(EMPIK.startUrl(), sourceScrapper).parallelStream()
```
Comment (KPP13) [inline src/main/java/io/github/bookcrawler/core/impl/EmpikBooksExtractor.java:22]: streams :heart:

File: src/main/java/io/github/bookcrawler/core/impl/SourceScrappingResult.java [hunk 4]
```diff
@@ -1,24 +1,21 @@
...
 
-
     SourceScrappingResult(Document source, SourceScrappingStatus status) {
         this.source = source;
         this.status = status;
     }
 
-    //TODO: not used method
-    public boolean isSuccesful() {
+    public boolean isSuccessful() {
```
Comment (KPP13) [inline src/main/java/io/github/bookcrawler/core/impl/SourceScrappingResult.java]: public is not necessary in here

File: src/main/java/io/github/bookcrawler/core/impl/SourceScrappingResult.java [hunk 5]
```diff
@@ -1,24 +1,21 @@
...
         return status == SourceScrappingStatus.SUCCESS;
     }
 
-    public Element getSource() {
+    public Document getSource() {
```
Comment (KPP13) [inline src/main/java/io/github/bookcrawler/core/impl/SourceScrappingResult.java]: public again :-1:

File: src/main/java/io/github/bookcrawler/core/impl/EmpikBookLinksCrawler.java [hunk 6]
```diff
@@ -2,42 +2,42 @@
...
-        this.sourceScrapper = sourceScrapper;
-    }
 
     @Override
-    public Collection<String> crawl() {
+    public Collection<String> crawl(String url, SourceScrapper sourceScrapper) {
+
+        SourceScrappingResult sourceScrappingResult = sourceScrapper.scrap(url);
+        if (!sourceScrappingResult.isSuccessful()) {
+            return Collections.EMPTY_LIST;
```
Comment (KPP13) [inline src/main/java/io/github/bookcrawler/core/impl/EmpikBookLinksCrawler.java]: :-1: please, use Collections.emptyList()

**Comment by KPP13** [comment]
Some remarks:
- you should add JavaDocs, at least before each interface and its public methods to explain the idea and behavior of the interface
- getters, setters, builders can be replaced with lombok and its annotations (if you don't want to use this then everything is ok now :) )
- logger should be added in some places (i.e. JsoupSourceScrapper - catch{} block), you can use it with an annotation before your class
- DiscountFetchingController:27 - you can use method reference instead of lambda - it will be easier to read this piece of code

I really like the idea of using @Override and abstract methods inside the enum :+1:

Other pluses and remarks are as comments in your pull request.

In summary - quite good, but according to me you need to make small changes before merging.
